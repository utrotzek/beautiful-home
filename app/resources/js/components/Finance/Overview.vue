<template>
    <div class="financial-overview">
        <div class="bank-balance">
            <span class="label">Kontostand:</span>
            <span class="amount positive">{{ accountingSum | toCurrency }}</span>
        </div>
        <hr >
        <h3>Umsätze</h3>
        <table
            id="currentAmountTable"
            class="table table-sm"
        >
            <thead>
                <tr>
                    <th></th>
                    <th>Einnahmen</th>
                    <th>Ausgaben</th>
                    <th>Summe</th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td class="font-weight-bold">
                        Aktuell
                    </td>
                    <td class="text-right">{{ accountingIncome | toCurrency }}</td>
                    <td class="text-right">{{ accountingOutgoings | toCurrency }}</td>
                    <td class="text-right">{{ accountingSum | toCurrency }}</td>
                </tr>
                <tr>
                    <td class="font-weight-bold">
                        Planung
                    </td>
                    <td class="text-right">{{ planningIncome | toCurrency }}</td>
                    <td class="text-right">{{ planningOutgoings | toCurrency }}</td>
                    <td class="text-right">{{ planningSum | toCurrency }}</td>
                </tr>
                <tr>
                    <td class="font-weight-bold">
                        Gesamt
                    </td>
                    <td class="text-right">{{ overallIncome | toCurrency }}</td>
                    <td class="text-right">{{ overallOutgoings | toCurrency }}</td>
                    <td class="text-right">{{ overallSum | toCurrency }}</td>
                </tr>
            </tbody>
        </table>

        <h3>Kostenstellen</h3>
        <table
            id="expensiveCostCenter"
            class="table table-sm"
        >
            <thead>
                <tr>
                    <th>Kostenstelle</th>
                    <th class="text-right">Summe</th>
                </tr>
            </thead>
            <tbody>
                <tr
                    v-for="(item, key) in topCostCenter"
                    :key="key"
                >
                    <td>{{ item.title }}</td>
                    <td class="text-right">{{ item.amount | toCurrency }}</td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script>
export default {
    name: "OverviewVue",
    filters: {
        toCurrency (value) {
            if (typeof value !== "number") {
                return value;
            }
            var formatter = new Intl.NumberFormat("de-DE", {
                style: "currency",
                currency: "EUR",
                minimumFractionDigits: 2
            });
            return formatter.format(value);
        }

    },
    props: {
        planningItems: {
            type: Array,
            required: true,
            default() {
                return [
                    {
                        id: 1,
                        costCenter: {
                            id: 1,
                            title: "Einkaufen"
                        },
                        description: "Gesamter Monat",
                        totalAmount: -100,
                        date: "25.10.2019",
                    }
                ];
            }
        },
        accountingItems: {
            type: Array,
            required: true,
            default() {
                return  [
                    {
                        id: 1,
                        title: "Westdeutsche Lotterie GmbH & Co. OHG westlotto.de",
                        totalAmount: 100,
                        remainingAmount: 0,
                        date: "12.10.2019",
                        display: true,
                        connectedPlanning: [
                            {
                                id: 10,
                                totalAmount: -100,
                                costCenter: {
                                    id: 6,
                                    title: "Lotto"
                                },
                                description: "Regeömäßiger Spielschein",
                            },
                        ],
                    }
                ];
            }
        }
    },
    computed: {
        accountingIncome() {
            let i=0;
            let sum=0;

            for (i=0; i < this.accountingItems.length; i ++) {
                let totalAmount = this.accountingItems[i].totalAmount;
                if (totalAmount > 0){
                    sum+=totalAmount;
                }
            }
            return sum;
        },
        accountingOutgoings() {
            let i=0;
            let sum=0;

            for (i=0; i < this.accountingItems.length; i ++) {
                let totalAmount = this.accountingItems[i].totalAmount;
                if (totalAmount < 0){
                    sum+=parseFloat(totalAmount);
                }
            }
            return sum;
        },
        accountingSum() {
            return this.accountingIncome + this.accountingOutgoings;
        },
        planningIncome() {
            let i=0;
            let sum=0;

            for (i=0; i < this.planningItems.length; i ++) {
                let totalAmount = this.planningItems[i].totalAmount;
                if (totalAmount > 0){
                    sum+=parseFloat(totalAmount);
                }
            }
            return sum;
        },
        planningOutgoings() {
            let i=0;
            let sum=0;

            for (i=0; i < this.planningItems.length; i ++) {
                let totalAmount = this.planningItems[i].totalAmount;
                if (totalAmount < 0){
                    sum+=parseFloat(totalAmount);
                }
            }
            return sum;
        },
        planningSum() {
            return this.planningIncome + this.planningOutgoings;
        },
        overallIncome() {
            return this.accountingIncome + this.planningIncome;
        },
        overallOutgoings() {
            return this.accountingOutgoings + this.planningOutgoings;
        },
        overallSum() {
            return this.accountingSum + this.planningSum;
        },
        topCostCenter() {
            let accIndex = 0;
            let topCostCenters = [];

            for (accIndex = 0; accIndex < this.accountingItems.length; accIndex++){
                let accountingItem = this.accountingItems[accIndex];
                let accConnectedIndex = 0;
                for (accConnectedIndex = 0; accConnectedIndex < accountingItem.connectedPlanning.length; accConnectedIndex++){
                    let connectedItem = accountingItem.connectedPlanning[accConnectedIndex];
                    let costCenterItem = connectedItem.costCenter;
                    let value = connectedItem.totalAmount;

                    if (value < 0){
                        value = value *-1;
                    }

                    let existingCostCenterItem = topCostCenters.find(costCenter => costCenter.id === costCenterItem.id );
                    if (!existingCostCenterItem){
                        topCostCenters.push({
                            title: costCenterItem.title,
                            amount: parseFloat(connectedItem.totalAmount),
                            value: parseFloat(value),
                            id: costCenterItem.id
                        });
                    }else{
                        existingCostCenterItem.amount += parseFloat(connectedItem.totalAmount);
                        existingCostCenterItem.value += parseFloat(value);
                    }
                }
            }

            topCostCenters = _.sortBy(topCostCenters, "value").reverse();
            return topCostCenters;
        }
    }
};
</script>

<style scoped>

</style>
