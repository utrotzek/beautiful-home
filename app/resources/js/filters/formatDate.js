import moment from "moment";

//TODO: set format in moment
export default function formatDate (value) {
    if (value) {
        if (typeof value === "object"){
            return moment(value.date).format("DD.MM.YYYY");
        }else{
            return moment(String(value)).format("DD.MM.YYYY");
        }
    }
}
