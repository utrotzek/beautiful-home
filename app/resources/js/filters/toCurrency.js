export default function toCurrency (value) {
    if (typeof value !== "number") {
        return value;
    }
    var formatter = new Intl.NumberFormat("de-DE", {
        style: "currency",
        currency: "EUR",
        minimumFractionDigits: 2
    });
    return formatter.format(value);
}
