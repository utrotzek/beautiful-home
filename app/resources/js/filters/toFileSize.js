export default function toFileSize(value){
    if (isNaN(value)){
        return "undefined size";
    }else{
        const units = [ "byte", "kB", "MB", "GB", "TB"];
        let newValue = value;
        let counter = 0;

        if (newValue >= 1000){
            while (newValue >= 1000){
                newValue = newValue / 1024;
                counter++;
            }
        }

        var formatter = new Intl.NumberFormat("de-DE", {
            minimumFractionDigits: 1,
            maximumFractionDigits: 1
        });
        return formatter.format(newValue) + " " + units[counter];
    }
}
