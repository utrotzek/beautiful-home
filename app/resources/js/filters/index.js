import toCurrency from "./toCurrency";
import formatDate from "./formatDate";
import toFileSize from "./toFileSize";

export default {
    install(Vue) {
        Vue.filter("toCurrency", toCurrency);
        Vue.filter("formatDate", formatDate);
        Vue.filter("toFileSize", toFileSize);
    }
};
