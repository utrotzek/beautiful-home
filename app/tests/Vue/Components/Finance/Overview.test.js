import {mount} from "@vue/test-utils";
import Vue from "vue";
import Overview from "../../../../resources/js/components/Finance/Overview";
import filters from "../../../../resources/js/filters";
Vue.use(filters);

import AccountingFixtures from "./Fixtures/AccountingFixtures";
import PlanningFixtures from "./Fixtures/PlanningFixtures";

describe("Overview", () => {
    let wrapper = null;
    const planningItems = PlanningFixtures;
    const accountingItems = AccountingFixtures;

    beforeEach(() => {
        wrapper = mount(Overview,{
            propsData: {
                accountingItems: accountingItems,
                planningItems: planningItems
            }
        });
    });

    it ("renders properly", () => {
        expect(wrapper.element).toMatchSnapshot();
    });


    it ("calculates accountingIncome correctly", () => {
        expect(wrapper.vm.accountingIncome).toBe(1100.0);
    });

    it ("calculates accountingOutgoings correctly", () => {
        expect(wrapper.vm.accountingOutgoings).toBe(-150.0);
    });

    it ("calculates accountingSum correctly", () => {
        expect(wrapper.vm.accountingSum).toBe(950.0);
    });

    it ("calculates planningIncome correctly", () => {
        expect(wrapper.vm.planningIncome).toBe(1100.0);
    });

    it ("calculates planningOutgoings correctly", () => {
        expect(wrapper.vm.planningOutgoings).toBe(-250.0);
    });

    it ("calculates planningSum correctly", () => {
        expect(wrapper.vm.planningSum).toBe(850.0);
    });

    it ("calculates overallIncome correctly", () => {
        expect(wrapper.vm.overallIncome).toBe(2200.0);
    });

    it ("calculates overallOutgoings correctly", () => {
        expect(wrapper.vm.overallOutgoings).toBe(-400.0);
    });

    it ("calculates overallSum correctly", () => {
        expect(wrapper.vm.overallSum).toBe(1800.0);
    });

    it ("calculates topIncomeCostCenter correctly", () => {
        const expectedIncomeArray = [
            {
                title: "Gehalt",
                amount: 700,
                value: 700,
                id: 10
            },
            {
                title: "Lotto",
                amount: 400,
                value: 400,
                id: 11
            }
        ];
        expect(wrapper.vm.topIncomeCostCenter).toEqual(expectedIncomeArray);
    });

    it ("calculates topOutgoingCostCenter correctly", () => {
        const expectedIncomeArray = [
            {
                title: "Einkaufen",
                amount: -90,
                value: 90,
                id: 1
            },
            {
                title: "Sonstiges",
                amount: -60,
                value: 60,
                id: 3
            }
        ];
        expect(wrapper.vm.topOutgoingCostCenter).toEqual(expectedIncomeArray);
    });

    it ("can filter cost centers", () => {
        wrapper.find("#top-outgoing tr:nth-child(1) .filter").trigger("click");

        const expectedOneFilter = [
            {
                id: 1,
                title: "Einkaufen"
            },
        ];
        expect(wrapper.emitted().filterChanged[0]).toEqual([expectedOneFilter]);

        wrapper.find("#top-outgoing tr:nth-child(2) .filter").trigger("click");

        const expectedTwoFilter = [
            {
                id: 1,
                title: "Einkaufen"
            },
            {
                id: 3,
                title: "Sonstiges"
            },
        ];
        expect(wrapper.emitted().filterChanged[1]).toEqual([expectedTwoFilter]);
    });

    it ("can unfilter cost centers", () => {
        wrapper.find("#top-outgoing tr:nth-child(1) .filter").trigger("click");
        wrapper.find("#top-outgoing tr:nth-child(2) .filter").trigger("click");
        wrapper.find("#top-outgoing tr:nth-child(1) .filter").trigger("click");
        wrapper.find("#top-outgoing tr:nth-child(2) .filter").trigger("click");

        const expectedNoFilter = [];
        expect(wrapper.emitted().filterChanged[0]).toEqual([expectedNoFilter]);
    });
});
