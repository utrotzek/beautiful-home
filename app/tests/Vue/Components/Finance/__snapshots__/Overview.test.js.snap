// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview renders properly 1`] = `
<div
  class="financial-overview"
>
  <div
    class="bank-balance"
  >
    <span
      class="label"
    >
      Kontostand:
    </span>
     
    <span
      class="amount positive"
    >
      950,00 €
    </span>
  </div>
   
  <table
    class="table table-sm mt-3"
    id="currentAmountTable"
  >
    <thead>
      <tr>
        <th
          class="text-right"
        />
         
        <th
          class="text-right"
        >
          
                    Einnahmen
                
        </th>
         
        <th
          class="text-right"
        >
          
                    Ausgaben
                
        </th>
         
        <th
          class="text-right"
        >
          
                    Summe
                
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr>
        <td
          class="font-weight-bold"
        >
          
                    Umsätze
                
        </td>
         
        <td
          class="text-right"
        >
          
                    1.100,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    -150,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    950,00 €
                
        </td>
      </tr>
       
      <tr>
        <td
          class="font-weight-bold"
        >
          
                    Planung
                
        </td>
         
        <td
          class="text-right"
        >
          
                    1.100,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    -250,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    850,00 €
                
        </td>
      </tr>
       
      <tr>
        <td
          class="font-weight-bold"
        >
          
                    Gesamt
                
        </td>
         
        <td
          class="text-right"
        >
          
                    2.200,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    -400,00 €
                
        </td>
         
        <td
          class="text-right"
        >
          
                    1.800,00 €
                
        </td>
      </tr>
    </tbody>
  </table>
   
  <div
    id="top-outgoing"
  >
    <h3>
      Ausgaben
    </h3>
     
    <table
      class="table table-sm"
    >
      <thead>
        <tr>
          <th>
            Kostenstelle
          </th>
           
          <th
            class="text-right"
          >
            
                        Summe
                    
          </th>
        </tr>
      </thead>
       
      <tbody>
        <tr>
          <td>
            <span
              class="filter fas fa-filter inactive"
            />
            
                        Einkaufen
                    
          </td>
           
          <td
            class="text-right"
          >
            
                        -90,00 €
                    
          </td>
        </tr>
        <tr>
          <td>
            <span
              class="filter fas fa-filter inactive"
            />
            
                        Sonstiges
                    
          </td>
           
          <td
            class="text-right"
          >
            
                        -60,00 €
                    
          </td>
        </tr>
      </tbody>
    </table>
  </div>
   
  <div
    id="top-income"
  >
    <h3>
      Einnahmen
    </h3>
     
    <table
      class="table table-sm"
    >
      <thead>
        <tr>
          <th>
            Kostenstelle
          </th>
           
          <th
            class="text-right"
          >
            
                        Summe
                    
          </th>
        </tr>
      </thead>
       
      <tbody>
        <tr>
          <td>
            <span
              class="filter fas fa-filter inactive"
            />
            
                        Gehalt
                    
          </td>
           
          <td
            class="text-right"
          >
            
                        700,00 €
                    
          </td>
        </tr>
        <tr>
          <td>
            <span
              class="filter fas fa-filter inactive"
            />
            
                        Lotto
                    
          </td>
           
          <td
            class="text-right"
          >
            
                        400,00 €
                    
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
